client/relay-chain-network/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+		self.rpc_client.parachain_host_session_index_for_child(hash).await
+	}
+
+	async fn validators(&self, block_id: &PHash) -> RelayChainResult<Vec<ValidatorId>> {
+		self.rpc_client.parachain_host_validators(block_id).await
+	}
+
+	async fn import_notification_stream(
+		&self,
+	) -> RelayChainResult<Pin<Box<dyn Stream<Item = PHeader> + Send>>> {
+		let imported_headers_stream =
+			self.rpc_client.subscribe_all_heads().await?.filter_map(|item| async move {
+				item.map_err(|err| {
+					tracing::error!(
+						target: LOG_TARGET,
+						"Encountered error in import notification stream: {}",
+						err
+					)
+				})
+				.ok()
+			});
+
+		Ok(Box::pin(imported_headers_stream))
+	}
+
+	async fn finality_notification_stream(
+		&self,
+	) -> RelayChainResult<Pin<Box<dyn Stream<Item = PHeader> + Send>>> {
+		let imported_headers_stream = self
+			.rpc_client
+			.subscribe_finalized_heads()
+			.await?
+			.filter_map(|item| async move {
+				item.map_err(|err| {
+					tracing::error!(
+						target: LOG_TARGET,
+						"Encountered error in finality notification stream: {}",
+						err
+					)
+				})
+				.ok()
+			});
+
+		Ok(Box::pin(imported_headers_stream))
+	}
+
+	async fn best_block_hash(&self) -> RelayChainResult<PHash> {
+		self.rpc_client.chain_get_head().await
+	}
+
+	async fn is_major_syncing(&self) -> RelayChainResult<bool> {
+		self.rpc_client.system_health().await.map(|h| h.is_syncing)
+	}
+
+	fn overseer_handle(&self) -> RelayChainResult<Option<Handle>> {
+		unimplemented!("Overseer handle is not available on relay-chain-network");
+	}
+
+	async fn get_storage_by_key(
+		&self,
+		relay_parent: &PHash,
+		key: &[u8],
+	) -> RelayChainResult<Option<StorageValue>> {
+		let storage_key = StorageKey(key.to_vec());
+		self.rpc_client
+			.state_get_storage(storage_key, Some(*relay_parent))
+			.await
+			.map(|storage_data| storage_data.map(|sv| sv.0))
+	}
+
+	async fn prove_read(
+		&self,
+		relay_parent: &PHash,
+		relevant_keys: &Vec<Vec<u8>>,
+	) -> RelayChainResult<StorageProof> {
+		let cloned = relevant_keys.clone();
+		let storage_keys: Vec<StorageKey> = cloned.into_iter().map(StorageKey).collect();
+
+		self.rpc_client
+			.state_get_read_proof(storage_keys, Some(*relay_parent))
+			.await
+			.map(|read_proof| {
+				let bytes = read_proof.proof.into_iter().map(|bytes| bytes.to_vec()).collect();
+				StorageProof::new(bytes)
+			})
+	}
+
+	/// Wait for a given relay chain block
+	///
+	/// The hash of the block to wait for is passed. We wait for the block to arrive or return after a timeout.
+	///
+	/// Implementation:
+	/// 1. Register a listener to all new blocks.
+	/// 2. Check if the block is already in chain. If yes, succeed early.
+	/// 3. Wait for the block to be imported via subscription.
+	/// 4. If timeout is reached, we return an error.
+	async fn wait_for_block(&self, wait_for_hash: PHash) -> RelayChainResult<()> {
+		let mut head_stream = self.rpc_client.subscribe_all_heads().await?;
```

Comment:
Do you know when this future returns here? Does it return when the other side has signaled us that the subscription is done?